Use track 2 code to set up Search live tests

heaths opened the pull request.

heaths committed:
Use track 2 code to set up Search live tests

heaths commented:
Once the other PR is merged, I'll rebase this and any future commits back onto master so it's just these changes you see now. This also fixes the live tests.

heaths commented:
/azp run net - search - ci

tg-msft commented on sdk/search/Azure.Search.Documents/tests/Utilities/SearchResources.Data.cs:
```diff
@@ -32,48 +23,51 @@ public partial class SearchResources
-                    Field.New("rating", DataType.Int32, isFilterable: true, isSortable: true, isFacetable: true),
-                    Field.New("location", DataType.GeographyPoint, isFilterable: true, isSortable: true),
-                    Field.NewComplex("address", isCollection: false, fields: new[]
+                    new SimpleField("hotelId", DataType.String) { IsKey = true, IsFilterable = true, IsSortable = true, IsFacetable = true },
```
This whole section looks really nice...

tg-msft commented on sdk/search/Azure.Search.Documents/tests/DocumentOperations/SearchTests.cs:
```diff
@@ -116,6 +116,7 @@ public async Task<SearchResources> CreateLargeHotelsIndexAsync(int size)
 
         [Test]
+        [Ignore("Complex fields are not being serialized: https://github.com/Azure/azure-sdk-for-net/issues/10944")]
         public async Task DynamicDocuments()
```
Thanks for digging through this

tg-msft commented on sdk/search/Azure.Search.Documents/tests/Utilities/SearchResources.cs:
```diff
@@ -408,16 +409,11 @@ private async Task<SearchResources> CreateSearchServiceAndIndexAsync()
 
                 // Give the index time to stabilize before running tests.
-                await Task.Delay(TimeSpan.FromSeconds(20));
+                await WaitForIndexingAsync();
```
I don't think this is the same, is it?  These were tuned the hard way by Bruce and I'd love to keep them if there isn't a good reason to change them.

heaths replied:
There are other helps, but all that does is no-op if using a recording, or if a delay was specified. If we're merely playing back a test, we shouldn't waste time waiting. Maybe I'll just have a generic conditional waiter to avoid confusion about why we're waiting.

tg-msft replied:
We don't run this code path when playing back anything.

heaths replied:
Fair point, but 20 seconds? IMO, the service needs to use LROs in more places. This magic number is probably a safe bet, but means then it's too long in the typical case. All tests passed with the method I used, which, IIRC, was only 2 seconds. Do we really need to add another 18 seconds?

tg-msft replied:
Bruce said he spent a lot of time and energy getting this exact behavior dialed in and I'd prefer to stick with it if we can.

heaths replied:
I'll make it configurable. All 400+ live tests passed yesterday with only a 2 second delay. I'll default to 20 as noted here. /cc @brjohnstmsft

tg-msft commented on sdk/search/Azure.Search.Documents/tests/Utilities/SearchResources.cs:
```diff
@@ -224,7 +225,7 @@ public static async Task<SearchResources> CreateWithHotelsIndexAsync(SearchTestB
         public static async Task<SearchResources> GetSharedHotelsIndexAsync(SearchTestBase fixture)
         {
-            SharedSearchResources.Search ??= await CreateWithHotelsIndexAsync(fixture);
+            await SharedSearchResources.EnsureInitialized(async () => await CreateWithHotelsIndexAsync(fixture));
```
Awesome

tg-msft commented on sdk/search/Azure.Search.Documents/tests/Utilities/SharedSearchResources.cs:
```diff
@@ -14,10 +15,38 @@ namespace Azure.Search.Documents.Tests
+        /// </summary>
+        /// <param name="factory">The method to create the shared search index if necessary.</param>
+        /// <returns>A <see cref="ValueTask"/> that the method has completed.</returns>
+        public static async ValueTask EnsureInitialized(Func<Task<SearchResources>> factory)
```
If you had the same problem with KeyVault tests, should we pull this into the shared source testing code at some point?

heaths replied:
The only problem was people using `??=` where thread safety was needed. No different than using `+=` etc. It's just something people need to avoid. I wouldn't be opposed to adding this to shared code, but I know @pakrym generally wants to avoid hard-to-discover helpers people may not use. But if you think this is worth it, we can make the case.

The fix for KV wasn't like this. It was simpler than that because the impact of overwriting a shared variable wasn't nearly as bad - it was just an in-memory cache of challenge scopes so we avoid lots of string parsing / allocations.

tg-msft replied:
I see - I thought it also involved the logic in this method.  That's fair then.

tg-msft reviewed (approved):
This looks awesome - thanks for making this change.

pakrym commented on sdk/search/Azure.Search.Documents/tests/Utilities/SharedSearchResources.cs:
```diff
@@ -14,10 +15,38 @@ namespace Azure.Search.Documents.Tests
+        /// </summary>
+        /// <param name="factory">The method to create the shared search index if necessary.</param>
+        /// <returns>A <see cref="ValueTask"/> that the method has completed.</returns>
+        public static async ValueTask EnsureInitialized(Func<Task<SearchResources>> factory)
```
How would this work when two tests pass different initialization functions? Who wins?

heaths replied:
Doesn't really matter since the method is `EnsureInitialized`, using the same pattern/behavior as `LazyInitializer.EnsureInitialized`. The only winner is whoever sets the `Search` property first. How or what they do doesn't matter.

pakrym replied:
It does matter because `LazyInitializer.EnsureInitialized` also takes a destination while you don't. So if two callers pass different lambdas that intend to create different resources they'll trample over each other.

heaths replied:
Fair point, but this is to initialize this particular shared instance of a class that would otherwise be static if not for supporting instantiation to support NUnit. This isn't a generic use case. The idea was to be a simple call so callers *don't* have to think about how to call this.

pakrym replied:
Then why don't you move CreateWithHotelsIndexAsync method into SharedSearchResources.cs and expose SearchResources via a method like ValueTask<SearchResources> GetSearchResourcesAsync()

heaths replied:
Because it's also used outside that context as well. I don't see any value to change this as it's isolated for a very particular use case. It's target is very specific, and your argument would also apply if someone used `LazyInitializer.EnsureInitialized` with the same `ref` argument but different factory method as well.

heaths committed:
Resolve PR feedback

heaths commented:
Force pushed to rebase onto master.

heaths commented:
/azp run net - search -ci

heaths commented:
/azp run

heaths commented:
/azp run net - search - ci

heaths commented:
I can't seem to get rid of the failed status checks before I rebased the PR (after I rebased the branch and pushed). Going to close and reopen a new PR.

heaths commented:
Failed checks are associated with the commit, not the PR, so opening a new one won't help. I'm going to try rebasing and force push again to hopefully clear these other ones out. (GitHub needs to support `git rebase --onto` flows.)
